### PR TITLE
For #11404 : "Open all..." option

### LINKED ISSUE #11404: "Open all..." option (Synced tabs, Bookmark Folder, etc)
### What is the user problem or growth opportunity you want to see solved?
Users should be able to open all of their synced tabs in Fenix in order to quickly restore a session from another device.

### How do you know that this problem exists today? Why is this important?
My Pixel is getting replaced due to a hardware issue. I want to maintain my user data from my previous device, and https://github.com/mozilla-mobile/fenix/issues/2080 is not yet a feature. Google restores also don't restore my session from the cloud.

The closest feature to restore my tabs from my old device is sync, but it is extremely cumbersome to tap 3 times for every tab I want to migrate. I have over 100 tabs in my old device, so that at a minimum, 300 taps on my screen along with the sheer inefficiency of the process. 

### Who will benefit from it?
Anyone who wants to migrate their existing session to a new device.

┆Issue is synchronized with this [Jira Task](https://mozilla-hub.atlassian.net/browse/FNXV2-1892)

### PR DESCRIPTION
**Pull request by Taknok**
Add the "Open all..." option as described in #11404 (similar to "Open all" present on firefox desktop).

This option open recursively all link in a bookmark folder, if there is no link in the folder, display a toast message with the error. All strings a retrieved from "l10n-mozilla firefox-desktop". Once links are opened, the tabs tray is shown.

Folder structure for this exemple:
Screenshot_1631196565

How the option is shown in overflow menu:
Screenshot_1631196590

When opening all:
Screenshot_1631196598

When a folder is empty:
Screenshot_1631196580

### Pull Request checklist

- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [x] **Screenshots**: This PR includes screenshots or GIFs of the changes made or an explanation of why it does not
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features. In addition, it includes a screenshot of a successful [accessibility scan](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor&hl=en_US) to ensure no new defects are added to the product.

### To download an APK when reviewing a PR:
1. click on Show All Checks,
2. click Details next to "Taskcluster (pull_request)" after it appears and then finishes with a green checkmark,
3. click on the "Fenix - assemble" task, then click "Run Artifacts".
4. the APK links should be on the left side of the screen, named for each CPU architecture

Close #11404
Fixes #11404
Fixes #11404
Fixes #11404

### CODE REVIEW

**Comment by Taknok** [comment]
Can a reviewer tell me if this is a correct way to introduce new strings ? As both chats fenix an l10n in chat.mozilla.org couldn't tell me if I should push it 1st on l10n or fenix project.

**Comment by Mugurell** [comment]
> Can a reviewer tell me if this is a correct way to introduce new strings ? As both chats fenix an l10n in chat.mozilla.org couldn't tell me if I should push it 1st on l10n or fenix project.

All new string resources are to be added only in Fenix `values/strings.xml` - only to the default strings file. 
All localization is done automatically, we should not manually update localized strings in Fenix.

**Comment by Taknok** [comment]
Ok, so I should only set in `values/strings.xml` wait this string is imported in l10n, copy values on Pontoon from desktop to fenix and wait auto import here ?

**Comment by Mugurell** [comment]
> Ok, so I should only set in `values/strings.xml` wait this string is imported in l10n, copy values on Pontoon from desktop to fenix and wait auto import here ?

Once the PR is merged with the changes in `values/strings.xml` the work on Fenix is done.
If you want to help with translations on Pontoon you can certainly do that and would be highly appreciated but that work is outside of the scope of the Fenix ticket.

**Comment by gabrielluong** [comment]
We probably want to check with Product/UX before proceeding.

**Comment by Taknok** [comment]
I thought that the interface modification was not that important, but ok. Should I contact them, fill an issue/a form or they will pass on their own ?

**Comment by amedyne** [comment]
@topotropic or @vesta0 Any thoughts  on this? 
@Taknok FYI. tagging them here for input.

**Comment by Taknok** [comment]
According to [this documentation](https://github.com/mozilla-mobile/fenix/wiki/Firefox-for-Android-Team-Processes#ux-review) an engineer should contact the designer (desk or slack).

**Comment by Taknok** [comment]
Force-push to rebase PR onto main branch.

Any update about this UX review ?

**Comment by amedyne** [comment]
Talked to product about this PR, the request is to implement telemetry as part of this change. @gabrielluong or @Mugurell here's the Jira ticket: https://mozilla-hub.atlassian.net/browse/FNXV2-17873 for the change. Do you think you can take on this or mentor/guide @Taknok to implement?

**Comment by Taknok** [comment]
Thank you for your quick response. I cannot access the jira ticket, thus I cannot see what is asked. What should be implemented in the telemetry ? Create a special entry ? Count all tabs opened ? Just register one count even if multiples are opened ?

**Comment by amedyne** [comment]
> Thank you for your quick response. I cannot access the jira ticket, thus I cannot see what is asked. What should be implemented in the telemetry ? Create a special entry ? Count all tabs opened ? Just register one count even if multiples are opened ?

The request is to register one count if a user selects the "Open All Bookmarks" action.

**Comment by Taknok** [comment]
Can I link "Open all" from the overflow menu (this PR) to the event [`Event.OpenedBookmarksInNewTabs`](https://github.com/mozilla-mobile/fenix/blob/6ac10d520931b76e1d4db50ce0936a8e10aaa1d9/app/src/main/java/org/mozilla/fenix/components/metrics/GleanMetricsService.kt#L186) which is currently used when a user select multiple bookmark items and click open (which currently does not work with folders, this is why this issue and PR were created).

To sum up:
Event `Event.OpenedBookmarksInNewTabs` would be triggered when a user open a folder menu and click "Open all"  _**and/or**_ when user select multiple items (no folders), click tree dots menu from selection bar and click "open in a new bookmark".

**Comment by eliserichards** [comment]
I think we should make a new Event for open all, just so we can differentiate. What do you think @Taknok?

**Comment by eliserichards** [comment]
e.g. `Event.OpenAllBookmarksInNewTabs`

**Comment by Taknok** [comment]
Here is the metric with Glean.

**Comment by Taknok** [comment]
Force-push to rebase PR onto main branch.

Any update about this UX review ?

File: app/src/main/java/org/mozilla/fenix/library/bookmarks/BookmarkItemMenu.kt
Comment (Taknok) [reply app/src/main/java/org/mozilla/fenix/library/bookmarks/BookmarkItemMenu.kt:96]: If possible, I could check the browser state (in the normal mode, or private mode) and display the string and opening the tabs accordingly to the current state:
- user accessing the bookmarks from private tabs > open all in private tabs
- user accessing the bookmarks from normal tabs > open all in new tabs

File: app/src/main/java/org/mozilla/fenix/library/bookmarks/BookmarkFragment.kt
Comment (Mugurell) [reply app/src/main/java/org/mozilla/fenix/library/bookmarks/BookmarkFragment.kt:290]: Indeed, right now the `BookmarkNode` only contains children information for the current node.
If we are to ask for all nested children in the current node by using `true` [here](https://github.com/mozilla-mobile/fenix/blob/f3bf56ac600481ef8b6db7bea189f799a006b023/app/src/main/java/org/mozilla/fenix/library/bookmarks/BookmarkFragment.kt#L265) then we'll be able to check whether the folder is empty or not and know if to show that menu option or not.

Regarding the performance aspect it looks like we only need one more level of children information, not all of them (we'd recursively get info about all children by using `true` in `getTree` as recommended above) so maybe we could get a new api for just this - asking for any number of children info levels. - I'll look into this and hopefully you'll have a new api to use soon.
Alternatively we could add this info later and take advantage of the DiffUtil we already have to not cause jank in the list but just asking for one more level of children info from the start seems simpler.

File: app/src/main/java/org/mozilla/fenix/library/bookmarks/BookmarkItemMenu.kt
Comment (Mugurell) [reply app/src/main/java/org/mozilla/fenix/library/bookmarks/BookmarkItemMenu.kt:96]: For individual bookmarks there is no such check or limitation.
There are just two different options.
Whether we should show or not a different option for opening bookmarks in private tabs can be addressed in a later ticket and PR, it should not block the current.

File: app/src/main/res/values/strings.xml
Comment (betsymi) [reply app/src/main/res/values/strings.xml:955]: Yes, looks good! Confirming "Open all in new tabs" and "Open all in private tabs" are good. Thanks @Mugurell. And thanks for the capitalization check as well.

File: app/src/main/java/org/mozilla/fenix/library/bookmarks/BookmarkItemMenu.kt
Comment (Taknok) [reply app/src/main/java/org/mozilla/fenix/library/bookmarks/BookmarkItemMenu.kt]: Done in a88a0fa .

File: app/metrics.yaml
Comment (Taknok) [reply app/metrics.yaml]: Done in 69cee3b .

File: app/src/main/res/values/strings.xml
Comment (Taknok) [reply app/src/main/res/values/strings.xml]: Done in d6bdc5d .

**Comment by Taknok** [comment]
Should I wait the implementation of api and resolution in `lib.rs` or goes for `bookmarks_get_tree` for now, and evolve to new api later ?

**Comment by Mugurell** [comment]
> Should I wait the implementation of api and resolution in `lib.rs` or goes for `bookmarks_get_tree` for now, and evolve to new api later ?
> 
> I'm afraid that `bookmarks_get_tree` could slow down a lot the display of the 3 dots menu when there is a lot of elements in a folder.

We can give it a try and see.
We are already using the `getTree` method with the `recursive` parameter set to true in `HomeActivity` so it would probably be okay to use it here also.
This would unblock showing the new menu option(s) only when they are appropriate - when the bookmark folder actually has bookmarks that can be opened and simplify the code a bit but avoiding that callback for a snackbar.

Thank you for pushing through with this!

**Comment by Taknok** [comment]
> We are already using the getTree method with the recursive parameter set to true in HomeActivity so it would probably be okay to use it here also.

Ok, I will try and pushed if convinced.

> This would unblock showing the new menu option(s) only when they are appropriate - when the bookmark folder actually has bookmarks that can be opened and simplify the code a bit but avoiding that callback for a snackbar.

Ok.

> One small thing regarding the patch: please try to squash the commits to standalone ones, [...]

I'm not sure to understand what you expect. Can you precise a quick example ? I tried todo micro commit as requested in `contribution.md`. Do you want something like : commit implementation, commit tests & commit androidtests ?

**Comment by Mugurell** [comment]
> I'm not sure to understand what you expect. Can you precise a quick example ? I tried todo micro commit as requested in `contribution.md`. Do you want something like : commit implementation, commit tests & commit androidtests ?

I'd say a commit should help solve a particular a particular flow, a particular unit of app functionality.
Looking through the list of current commits I see some that can be merged with the others to ensure cohesiveness such that a particular commit helps solve an entire particular scenario.
For example:
- "Add string to display in overflow menu" can be merged with the one where this string is used.
-  "Create core function" can be merged with the code changes that will actually call this new method.
- the work for telemetry can be similarly be a single commit, indeed separate from the functionality of opening the bookmark folder.
- tests should be part of the same commit which contains the code they test.
- static analysis fixups and addressing review comments don't need a separate commit. Ideally these should be merged with the initial commits that need those changes such that those commits will look right in the first place.

Nothing major, I know finding the right balance is hard and again appreciate your investment in this. It's great to separate unrelated functionality or break big changes in smaller units but having a commit for an entire piece of functionality also helps the reviewer and later code archaeologists understand how a feature works by having all it's needed changes in one commit, easy to reason about.

File: app/src/main/java/org/mozilla/fenix/library/bookmarks/BookmarkItemMenu.kt
Comment (Taknok) [reply app/src/main/java/org/mozilla/fenix/library/bookmarks/BookmarkItemMenu.kt]: Partially done in eb07f89 (this commit will be squash when approved).

I only check one level of children in the display or not check. I chose this, because we are forced to use a coroutine since `getTree` is async and on slow android with 10 levels and ~20 items was quite bugging because coroutine has not yet finished. Keeping one level seems me not that bad. (I will fully check all level when user has clicked on open all to show a war message if >15 as requested in others reviews.)

So:
```
root
 |-> folder1
 |    | -> item1
 |    |-> folder2
 |         | -> item2
 | -> emptyfolder1
       | -> emptyfolder2
```
folder1, folder2 & emptyfolder1 show option. Not emptyfolder2.

Comment (Mugurell) [reply app/src/main/java/org/mozilla/fenix/library/bookmarks/BookmarkItemMenu.kt]: That is great! Thank you!

File: app/src/test/java/org/mozilla/fenix/library/bookmarks/BookmarkItemMenuTest.kt
Comment (Mugurell) [inline app/src/test/java/org/mozilla/fenix/library/bookmarks/BookmarkItemMenuTest.kt:51]: I see these tests failing after the last patch because of
```
io.mockk.MockKException: no answer found for: PlacesBookmarksStorage(#11).getTree
```

Seems like this can be overcome by using something like
```
    fun `delete item has special styling`() = runBlocking {
        var deleteItem: TextMenuCandidate? = null
        mockkStatic("org.mozilla.fenix.ext.BookmarkNodeKt") {
            every { any<Context>().bookmarkStorage } returns mockk(relaxed = true)

            deleteItem = menu.menuItems(BookmarkNodeType.SEPARATOR, "").last()
        }
```
later mocking even the `getTree` response.

(`menuItems` sounds like a property to me, seems a bit strange to receive arguments but will leave it up to you if you think it would also sound better and offer more context as maybe `createMenuItems()`)

**Comment by Mugurell** [comment]
Thank you for pushing through @Taknok !
Tested it a bit manually and it works beautifully, seems like just some tests need to be updated and then do a quick last review and look into merging.

Also adding Oana as a reviewer for the ui tests.

**Review by sv-ohorvath — APPROVED** [review]
I see no issue with the changes made to UI tests. Thanks for adding these.

**Comment by Taknok** [comment]
Alert message is implemented ! An alert message is displayed when more than 15 tabs are loaded. Now all requested modifications are implemented. I let you review PR.

By the way, there is no warning message for collections, even with a lot of tabs ;)

**Comment by Taknok** [comment]
Really T.T I merged 30min ago, and a PR 10min ago conflict T.T

**Comment by Mugurell** [comment]
@Taknok Again, thank you for pushing through with this patch!
It looks really nice!

One interesting approach is the usage of the `AlertDialog` for informing users about opening many tabs.
It's an important functionality for which we didn't have UX specs so could please show this on video to ask UX's opinion on how it all looks and works? 
One question I'd have for UX about this is whether the dialog should be cancelable when tapping outside of it or not.

(We could also leave this for a followup ticket later if it proves to still require changes because otherwise this patch looks good 👍)

**Comment by Taknok** [comment]
Hi,

As show in the 1st video you cannot tap outside to cancel the `AlertDialog`, but it is easily to change this behavior by setting true to this [line](https://github.com/Taknok/fenix/blob/ef50b20d36821ac286d0143e43c27d9cf400f8f3/app/src/main/java/org/mozilla/fenix/library/bookmarks/BookmarkFragment.kt#L305). Tell me if needed and I commit this ASAP.

Cancel with false:
https://user-images.githubusercontent.com/12498020/143613887-1d5d6601-feb9-4596-b96d-b012c53ea560.mp4

Cancel with true:
https://user-images.githubusercontent.com/12498020/143614569-40712187-021d-449c-8519-a98914bf6294.mp4

Accept:
https://user-images.githubusercontent.com/12498020/143614754-3118b4f6-bee8-44bb-a67b-be02e5b673e3.mp4

File: app/src/main/java/org/mozilla/fenix/library/bookmarks/BookmarkController.kt
Comment (Mugurell) [inline app/src/main/java/org/mozilla/fenix/library/bookmarks/BookmarkController.kt:76]: nit: Could this be marked as internal if it's not intended to be public?
Negligible improvement but it's a flow that we normally use for something that should be private but also want it available in tests.
This could even be a top level const.

File: app/metrics.yaml
Comment (Mugurell) [inline app/metrics.yaml:1927]: This will have to be updated.
Asking @eliserichards on whether this should probe should also be set to never expire.

File: app/metrics.yaml
Comment (Mugurell) [inline app/metrics.yaml:1979]: Same as the above, maybe this should never expire.

**Review by Mugurell — COMMENTED** [review]
Just a nit and question for Elise regarding the expiration date of the new probes.
Maybe they should be set to never expire, the same as the similar ones for individual bookmarks.
Reached out to UX which will evaluate the changes in the coming days

File: app/src/main/java/org/mozilla/fenix/library/bookmarks/BookmarkController.kt
Comment (Taknok) [reply app/src/main/java/org/mozilla/fenix/library/bookmarks/BookmarkController.kt:76]: It was marked as a top level constant during development, but for testing I needed to lower the constant, which is not possible (const), so I switched to a private else in test.
So should I set `internal` and also delete the `Visible for testing` ? Just `internal` ?

File: app/metrics.yaml
Comment (Taknok) [reply app/metrics.yaml:1927]: The expiration date was set to 1st December because it was set for others similar functions as in `open_in_private_tabs`. But #21038 done by @mdboom changes the expiration to `never`. I can also switch this to `never` if approved here (or a any desired value). Mean while tests will be failing since 1st is passed.

Comment (Mugurell) [reply app/metrics.yaml:1927]: Just to get a confirmation from Elise that it's okay to add her email here also.

**Comment by kbrosnan** [comment]
@Mugurell Assuming we take this we should file a followup ticket for guarding against too many tabs. Desktop warns on the user opening 15 or more tabs.

**Comment by Taknok** [comment]
I'm not sure to understand your comment, but the warning for 15 or more is already implemented in this PR.

File: app/src/main/java/org/mozilla/fenix/library/bookmarks/BookmarkController.kt
Comment (Mugurell) [reply app/src/main/java/org/mozilla/fenix/library/bookmarks/BookmarkController.kt:76]: Just setting this as `internal` and keeping `@VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)` sounds good, thank you!

**Comment by Mugurell** [comment]
> @Mugurell Assuming we take this we should file a followup ticket for guarding against too many tabs. Desktop warns on the user opening 15 or more tabs.

The code looks good but it indeed involves important changes so Nicole said will look through them and also involve the content designer about this prompt which would be copied to the app from desktop.
How it all will look can be seen in https://github.com/mozilla-mobile/fenix/pull/21212#issuecomment-980183263.

**Comment by Mugurell** [comment]
Thank you for rebasing this @Taknok !
Still waiting for @topotropic 's feedback regarding these changes.

**Comment by Taknok** [comment]
Push force:
- Rebase
- Fix MagicNumber
- Fix Strings

**Comment by Taknok** [comment]
Hi,

PR to fix conflict.

To be honest, I am quite tired to resolve merge conflict for 8 month without any advancement, seeing things that was requested in the past (like metrics) are now optional. Ok it is normal and part of the evolution of the projects, but taking my code over and over and rethinking to match new rules is tiring.

**Comment by Taknok** [comment]
image

**Comment by Taknok** [comment]
What am I missing here ? What do I need to do to make it accepted ? There is an issue linked and people wanting this basic functionality ... This stupid functionality is taking almost taking a year be be accepted...

**Comment by Mugurell** [comment]
Thank you @Taknok for pushing this. Asked again for feedback on merging the patch which is looking great!
One small issue which I saw is that if removing all bookmarks from a folder and stepping out of it the new options are still shown in the menu. This might be a bit tricky to solve now so we can also file a ticket to resolve after this would be merged.

**Comment by topotropic** [comment]
Thanks, @Taknok for your contribution! 

It would be great if you could update the warning that's displayed when someone tries to open more than 15 tabs to the following: 

> You are about to open 38 tabs. This might slow down Firefox while the pages are loading. Are you sure you want to continue?

--> so it's just about adding the "while the pages are loading" part. Appreciated!

**Comment by Taknok** [comment]
@Mugurell I agree for another ticket. Maybe you should mark que "requested changes" as resolved, I am not able to do that.

@topotropic Done.

**Comment by Mugurell** [comment]
> @Mugurell I agree for another ticket. Maybe you should mark que "requested changes" as resolved, I am not able to do that.

Sounds good. Will help land this and file the additional ticket when we get the final okay.

**Comment by issue-dispenser** [comment]
Hi, this is a feature that I'm quite looking forward to.

@Mugurell Any update on getting the "okay"? The issues seem to have been resolved.

**Review by Mugurell — APPROVED** [review]
Thank you!

**Comment by Mugurell** [comment]
@Taknok Two UI test are failing: `openAllInTabsTest` and `openAllInPrivateTabsTest` because of 
> androidx.test.espresso.NoMatchingViewException: No views in hierarchy found matching: view.getId() is <2131296673/org.mozilla.fenix.debug:id/close_bookmarks>

from where `}.closeMenu {` is called.

Can you please check and update?

I can also help with this, just let me know.

**Comment by Taknok** [comment]
Hi,
I will check in two days, for the PR anniversary.

**Comment by Taknok** [comment]
My code did not touch the `closeMenu` function. I have no issue with running it on my computer, but with `createFolder` of the ui test.

**Comment by Taknok** [comment]
So on my computer the function `createBookmark` is globally failing on first call (on main branch c7ebec360227c70ce2ba0243b88321eac39eee89) due to this popup changing the focus. 

Screenshot_20220910-011045__01


This popup shows when first navigating to an URL. This failure is present for all tests like in the `deleteMultipleSelectionTest` test which I didn't modified. So the tests involving al least this function are crashing a first time, then rerun (because the popup does not trigger anymore). Which is really a inefficient way of running tests.

**Comment by Mugurell** [comment]
@Taknok Good catch with the tests needing a rerun to pass.
Can you open a new ticket for that? (I can help also)
(the solution is simple, we just need to use `featureSettingsHelper.setTCPCFREnabled(false)`)

**Comment by Mugurell** [comment]
> My code did not touch the `closeMenu` function. I have no issue with running it on my computer, but with `createFolder` of the ui test.

To me it looks like we don't need that `closeMenu` call. There is no menu opened.
I see some new conflicts. We recently enabled a new `ktlint` rule which enforces trailing commas so the patch will have to be rebased and updated.
If you need help I can also take over to update and land this.

**Comment by Taknok** [comment]
> To me it looks like we don't need that `closeMenu` call. There is no menu opened.

There is a menu opened: the bookmarkmenu https://github.com/Taknok/fenix/blob/be8a7291b6e572acdf319e82ac65bc0f898d3877/app/src/androidTest/java/org/mozilla/fenix/ui/BookmarksTest.kt#L283. Plus this function is ok on my computer.


> I see some new conflicts. We recently enabled a new `ktlint` rule which enforces trailing commas so the patch will have to be rebased and updated.

I already merge 'successfully'  in local, but the `createBookmark` function fail in ui tests because it does not found the close button, which is present with an inspec tool... I didn't modify this code so I am not sure of what to do.

**Comment by Mugurell** [comment]
Will look into it.

**Comment by Mugurell** [comment]
@Taknok Rebased the commits to account for the new `ktlint` rule for enforcing trailing commas. Looked into updating the `openAllInTabsTest` & `openAllInPrivateTabsTest` tests but saw that since you need to interacts with the tabs tray - which is currently undergoing a refactoring to compose we'd better wait for the refactoring to complete so I `@Ignore`d these tests.

**Comment by Taknok** [comment]
All things passed ! Thank you !

**Comment by rocketsroger** [comment]
This landed without any data-review.  The probes are also set to never expire.  What happened here?

**Comment by rocketsroger** [comment]
Another issue is a string was updated directly.  A new string should have been added.  Since there are multiple issues, I'll have to revert this.  Let's land this with data-review and a expiry version.

**Comment by rocketsroger** [comment]
@Taknok unfortunately there are multiple issues here that we can't keep this in Nightly.  Please reopen this change with two updates:
1. Create a [data review ](https://github.com/mozilla/data-review/blob/main/request.md) for the telemetry ([example](https://github.com/mozilla-mobile/fenix/pull/26990#issuecomment-1247356997)) and tag the pr "need: data-review".  Make sure to not use "never" as expiry.
2. Please create a new string if a string update is unavoidable.  ([example](https://github.com/mozilla-mobile/fenix/pull/22052/commits/d881bc007d8376cc39b68daefc1cd2c9e632bc13))  

Please let me know if I can help.  Thanks

**Comment by Taknok** [comment]
Sorry but I will be rude here.

First:
> This landed without any data-review. The probes are also set to never expire. What happened here?
@rocketsroger 

I set an expiration date for probes in the initial PR, but I was told to change it to never here :
> Maybe they should be set to never expire
@Mugurell

Second:
> Another issue is a string was updated directly. A new string should have been added. Since there are multiple issues, I'll have to revert this. Let's land this with data-review and a expiry version.
@rocketsroger 

What string ? I only see line insert in this PR files app/src/main/res/values/strings.xml Do at least link the line with an issue.

Data-review ? Was not @eliserichards or @topotropic doing this ? I am so lost in this maze that this PR has been though...

Third:
This is totally not acceptable the way you all handle this. I am a benevolent contributor, but I feel handle like a infinite free disposable work force. I sorry but I am pissed off to the point I do not care if this is because Mozilla does not put enough energy/money/[what ever you want] in their mobile team. But I'm rebasing and resolving conflict on the same shitty basic feature for over a year !
I asked what should I do, I did everything that I was told. Finally it is accepted, and no, now I have to start over this hell ??? To do stuff initially done ??? So I am tired... so do it yourself, I am giving up.
This all thing convinced me to switch to Brave (I have been using Firefox since version 2 (2006 !)). Yes Firefox was one last alternative to big companies but Mozilla lost itself somewhere and I am profoundly sad about it.
Plus, now, on every approach on LinkedIn from Mozilla, I will simply reply a link of this PR !

**Comment by rocketsroger** [comment]
> ... But I'm rebasing and resolving conflict on the same shitty basic feature for over a year ! I asked what should I do, I did everything that I was told...

I understand your frustration.  I never expected you to know what's needed to land an issue and the reviewers should have given you proper instructions.  However, my hands are tied in this case.  At Mozilla we take privacy very seriously.  Any data collected have to go through data-review.  Any data that is collected without expiry needs product approval as well.  I want to help you re-land this.  Please put up a new PR when you're ready and I'll work with you (in a timely manner) to land this as soon as possible.  Thank you,

**Comment by rocketsroger** [comment]
@Taknok another solution is I can make the modifications and land this PR with you as a co-author.  I'll go through the data review for you as well.  If you're OK with my name associated with the PR that is.  Please let me know.  

Thank you for your passion and your contribution.

**Comment by Taknok** [comment]
You can do the modifications you want, I am just not doing it anymore, I will even accept PR from other contributors to my branch if needed.

**Comment by rocketsroger** [comment]
> You can do the modifications you want, I am just not doing it anymore, I will even accept PR from other contributors to my branch if needed.

I'll create a new PR with you as co-author.  Thanks

**Pull request merged**
